src/lib.rs

```diff
@@ -186,23 +198,54 @@ pub fn channel<T>() -> (Sender<T>, Receiver<T>) {
     // Allocate the channel on the heap and get the pointer.
     // The last endpoint of the channel to be alive is responsible for freeing the channel
     // and dropping any object that might have been written to it.
+
     let channel_ptr = Box::into_raw(Box::new(Channel::new()));
-    (Sender { channel_ptr }, Receiver { channel_ptr })
+
+    // SAFETY: `channel_ptr` came from a Box and thus is not null
+    let channel_ptr = unsafe { NonNull::new_unchecked(channel_ptr) };
+
+    (
+        Sender {
+            channel: channel_ptr,
+            _invariant: PhantomData,
+        },
+        Receiver {
+            channel: channel_ptr,
+        },
+    )
 }
 
 #[derive(Debug)]
 pub struct Sender<T> {
-    channel_ptr: *mut Channel<T>,
+    channel: NonNull<Channel<T>>,
```

Comment:
It is inconsistent to name the field `channel_ptr` in `SendError` but name it `channel` here when they currently are the exact same type. The inconsistency is spreading with a lot of code currently looking like `let channel_ptr = self.channel;`. I would love if that variable was named the same as the field. However, see my other comment on potentially changing the type in `SendError` :)

I think we should name the field `channel_ptr` if it contains a `NonNull<Channel<T>>`. And also because we need to have both `channel` (`&Channel`) and `channel_ptr` (`NonNull<Channel>`) in scope simultaneously in a lot of code. So it would be nice to be able to consistently distinguish between the two.

Reply:
Resolved!